### DISCUSSION TITLE
WebView with Toga doesn't run JavaScript on android but yes on Windows

### DISCUSSION THREAD

**Discussion by danielmedinaa** [beeware/toga#2173] [Debugging help]
I'm working on a WebView in python which should display a web page and it does it correctly, the issue is that in android the web page does not execute the JavaScript code that is on the page, also if I add one with invoke_javascript, this is only on android, on windows it does it without any problem.

**Comment by freakboy3742** [comment]
I presume this is on Toga 0.3.1? I can't say I've seen this myself, and in the main branch we've got tests that validate that Javascript is executing. Are you able to test this on the main branch? To do this, replace the reference to `toga-android==0.3.1` with:
```
requires = [
    "git+https://github.com/beeware/toga.git#subdirectory=core",
    "git+https://github.com/beeware/toga.git#subdirectory=android",
]
```

**Comment by danielmedinaa** [comment]
Thanks for replying!, I did that and it opens perfect but the JavaScript code still doesn't work, I'm using the apk directly, the I know that generates with briefcase build android, does that affect? Do I have to use Briefcase Package Android and use the AAB?

**Reply by freakboy3742** [reply]
That shouldn't make any difference - once the APK has been built, it's an Android binary; how you get it onto your phone doesn't matter.

Can you share anything about the specific Javascript you are running? Can you verify that *no* Javascript is running? Or is it just your Javascript specifically?

**Reply by mhsmith** [reply]
It would also be useful if you posted the relevant parts of your code.

**Reply by danielmedinaa** [reply]
Well, look, in test mode I added JavaScript code from the same python with invoke_javascirpt() and it's exactly the same, in the window that is generated with briefcase dev, and in the .exe everything works 100%, both the JavaScript that the page has and the simple alert that I made with the invoke_javascript,  however, neither the JavaScript of the page as such nor the JavaScript of inovke work
```
import toga
from toga.style.pack import CENTER, COLUMN, ROW, Pack


code = ''' alert('Hello') '''
class App(toga.App):
    def startup(self):
        self.main_window = toga.MainWindow(title=self.name)

        self.webview = toga.WebView(
            on_webview_load=self.on_webview_loaded, style=Pack(flex=1)
        )

        box = toga.Box(
            children=[
                toga.Box(
                    style=Pack(
                        direction=ROW,
                        alignment=CENTER,
                        padding=5,
                    ),
                ),
                self.webview,
            ],
            style=Pack(direction=COLUMN),
        )
        self.url = "http://nissi.cloud:7000" #Testing URL
        self.main_window.content = box
        self.webview.url = self.url

        # Show the main window
        self.main_window.show()

    def load_page(self, widget):
        self.webview.url = self.url

    def on_webview_loaded(self, widget):
        self.url = self.webview.url
        self.webview.invoke_javascript(code)


def main():
    return App("Nissi Cloud", "WebApp.Nissi.Apk")

def app_view():
    if __name__ == '__main__':
        main().main_loop()

app_view()
```
I've already taken into account the issue of http, in network_security_config I've already added our domain
```
<?xml version="1.0" encoding="utf-8"?>
<network-security-config>
    <domain-config cleartextTrafficPermitted="true">
        <domain includeSubdomains="true">127.0.0.1</domain>
        <domain includeSubdomains="true">localhost</domain>
        <domain includeSubdomains="true">nissi.cloud</domain>
    </domain-config>
</network-security-config>
```

**Reply by mhsmith** [reply]
I can't comment on the JavaScript in the page itself, because I don't know what it is. But the problem with `invoke_javascript` is probably that the `on_webview_loaded` method isn't even being called, because we don't support that event on Android yet. Sorry about that.

Also, it's not a good idea to invoke JavaScript immediately after setting the URL, because it might invoke on the previous page, or the new page in a partly-loaded state.

As a workaround, you could use a fixed delay:
```
    def startup(self):
        ...
        self.main_window.show()
        asyncio.create_task(self.task())

    async def task(self):
        await asyncio.sleep(1)
        self.webview.invoke_javascript(code)
```

**Reply by danielmedinaa** [reply]
Ok I understand thanks!, outside of the invoke, I noticed the bug that when I click on the button: "Add Image" a dialog box should come up that allows me to search for an image on the device since with JavaScript I call it. This function serves in the browser perfectly on all devices, and also in the .exe that I generate with briefcase but in the apk it does not work 

<html>
<body>


<form method="POST" action="http://nissi.cloud:7000/path-process-upload-image" enctype="multipart/form-data">
--
  | <input type="hidden" name="_token" value="ZyyB6JDkTnx6LmhWv6mffmS5ZVlY2SuCc4DP15Gv">                        <div class="d-flex">
  | <input type="button" class="btn btn-outline-primary btn-sm" value="Agregar Imagen" onclick ="javascript:document.getElementById('imageFile').click();">
  | <input id="imageFile" type="file" accept="image/jpg,png,jpeg" style='display:none;' name="product_image" required />
  | <input type="text" name="product_code" style='display:none;' value="code-product">
  | <button type="submit" class="btn btn-sm btn-outline-success ml-2">Guardar Imagen</button>
  | </div>
  | </form>


</body>
</html>

**Reply by mhsmith** [reply]
Is this a button inside a web page? Does the same problem happen when using that web page in a normal Android browser? If so, it's probably not caused by Toga, and you should search for help elsewhere.

**Reply by danielmedinaa** [reply]
The button works everywhere 100% even in toga but in windows, both running briefcase dev and in the exe, there everything works 100%, the problem is in the apk

**Reply by mhsmith** [reply]
OK, but you didn't answer either of the questions in my previous comment.

**Reply by danielmedinaa** [reply]
Is it a button within a web page?
  Answer: Yes. 
   Does the same problem occur when using that web page in a regular Android browser? 
   Answer: No

**Reply by mhsmith** [reply]
If you need help with that, please post the HTML of the page, or preferably a simple example that reproduces the same problem.

**Reply by danielmedinaa** [reply]
Exactly, I noticed the error here: 



that's the

**Comment by danielmedinaa** [comment]
[button.zip](https://github.com/beeware/toga/files/13284525/button.zip)

**Reply by mhsmith** [reply]
It looks like displaying a file picker from a WebView requires subclassing the WebChromeClient class. I'll add a note about this to #1020. Unfortunately I don't have any workaround at the moment.

**Reply by danielmedinaa** [reply]
Thank you!, another little thing and sorry for so many questions, is it possible to request permissions and runtime?? For example, android.permission.MANAGE_EXTERNAL_STORAGE?? Since a version of my APK needs to interact to read a txt file with the port, then to allow the service I have to do it manually and it is tedious for the users who are going to use it, then there is some way to request that permission but before the app opens

**Reply by freakboy3742** [reply]
Permissions like this need to be requested when building the app, so it's a feature we'd need in Briefcase, rather than Toga. At present, we don't have any mechanism to add permissions, but this is on our roadmap (see beeware/briefcase#547).

**Reply by mhsmith** [reply]
For some example code, see https://github.com/beeware/briefcase/discussions/1497#discussioncomment-7443705.

As well as the runtime code, you'd also need to add the permission to AndroidManifest.xml.
